### ISSUE TITLE
How to use behind proxy?

### ISSUE THREAD

**Issue by lshz0088** [tus/tusd#467]
**Question**

I set these parameters in tusd.

```
["--s3-bucket","bucketname","--s3-endpoint","http://minio-server:9000","--base-path","/bsh/oss/files/","--expose-metrics","false","--behind-proxy","true"]
```

in nginx
```
proxy_set_header Host $host:$server_port;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    # proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Proto https;
    proxy_request_buffering off;
    proxy_buffering off;
    proxy_http_version 1.1;
    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    client_max_body_size 0;
```

In actual production, there is an F5 agent in front of the nginx.

Now the phenomenon is:

The first request sent by the browser to save JSON was the HTTPS protocol, and it also responded. But the second request became http. Now Google is not allowed to send HTTP in HTTPS.

![image](https://user-images.githubusercontent.com/27722002/111965283-88846080-8b30-11eb-9c43-62e28fa024a8.png)
![image](https://user-images.githubusercontent.com/27722002/111965344-976b1300-8b30-11eb-9383-f648dfc73c8a.png)

**Comment by Acconut** [comment]
> In actual production, there is an F5 agent in front of the nginx.

What does that mean? Is there the F5 agent is use when you made these screenshots? If so, the F5 agent must be configured as well.

Also, there are multiple `proxy_set_header X-Forwarded-Proto` instructions in your config. Try to only keep the `proxy_set_header X-Forwarded-Proto https;` one.

**Comment by lshz0088** [comment]
The current situation is as follows:
1.  `F5` agent is configured by the third party enterprise.
2.  `SSL certificate` is configured on `F5` proxy
3.  There is no `SSL certificate` on my own nginx proxy, and the internal forwarding uses `http`
4.  `X-Forwarded-Proto` I also tried to write a `"HTTPS"`

> Tips:
> Actually, I haven't used `F5` myself, so I don't know what can be configured on `F5`. Do you know about `F5` agent?

**Comment by lshz0088** [comment]
@Acconut 
In addition, I want to confirm one thing.
Is the rule of the second HTTP request returned after the first request is sent?
Because I also found that if I remove the part of the location matching in the proxy process, the second request will also remove this part, which is different from the URL sent for the first time

**Comment by lshz0088** [comment]
@Acconut 
In addition, I want to confirm one thing.
Is the rule of the second HTTP request returned after the first request is sent?
Because I also found that if I remove the part of the location matching in the proxy process, the second request will also remove this part, which is different from the URL sent for the first time

**Comment by Acconut** [comment]
> Do you know about `F5` agent?

No, your description is the first time I hear about F5.

>  `X-Forwarded-Proto` I also tried to write a `"HTTPS"`

Did you also try with lowercase `https`?

> Is the rule of the second HTTP request returned after the first request is sent?
> Because I also found that if I remove the part of the location matching in the proxy process, the second request will also remove this part, which is different from the URL sent for the first time

Sorry, but I do not understand what you mean here. Can you rephrase this?

**Comment by lshz0088** [comment]
Is the whole upload process divided into two steps

1. Upload JSON with file information

2. Upload the actual file

Now, the first step is successful (I've seen the JSON file in Minio). It's a request for HTTPS.

But in the second step, it shows an HTTP request. Then it was rejected by the browser.

The endpoint I configured in tus JS clinet is:
https://www.xxx.com.cn/bsh/oss/files/

I've always configured lowercase HTTPS.

![image](https://user-images.githubusercontent.com/27722002/113978012-0016f700-9876-11eb-959a-6e98657e0626.png)

**Comment by lshz0088** [comment]
This is the first request. It carries a `Location`

![image](https://user-images.githubusercontent.com/27722002/113982458-5cc8e080-987b-11eb-9c63-a48c49e2b71c.png)

**Comment by lshz0088** [comment]
@Acconut The `Location` above is an HTTP protocol

**Comment by Acconut** [comment]
@lshz0088 There are still multiple `proxy_set_header X-Forwarded-Proto` instructions in your configuration. Please remove these conflicting instructions.

**Comment by sean9999** [comment]
I had this problem using and nginx-based kubernetes ingress proxy. The solution was to add something like this:

proxy_redirect http://mydomain.tld https://mydomain.tld;

http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_redirect

**Comment by Acconut** [comment]
Thank you for your input, @sean9999. That's very helpful if people come across this issue in the future.

Since I haven't heard back from the original author @lshz0088, I am going to close this issue. If the problem comes back, feel free to open a new ticket.
